tokenizer/token.go

```diff
@@ -0,0 +1,638 @@
+	TokenError:        &TokenExtraError{},
+	TokenBadEscape:    &TokenExtraError{},
+	TokenBadString:    &TokenExtraError{},
+	TokenBadURI:       &TokenExtraError{},
+	TokenHash:         &TokenExtraHash{},
+	TokenNumber:       &TokenExtraNumeric{},
+	TokenPercentage:   &TokenExtraNumeric{},
+	TokenDimension:    &TokenExtraNumeric{},
+	TokenUnicodeRange: &TokenExtraUnicodeRange{},
+}
+
+// TokenExtraHash is attached to TokenHash.
+type TokenExtraHash struct {
+	IsIdentifier bool
+}
+
+// Returns a descriptive string, either "unrestricted" or "id".
+func (e *TokenExtraHash) String() string {
+	if e == nil || !e.IsIdentifier {
+		return "unrestricted"
+	} else {
+		return "id"
+	}
+}
+
+// TokenExtraNumeric is attached to TokenNumber, TokenPercentage, and
+// TokenDimension.
+type TokenExtraNumeric struct {
+	// Value float64 // omitted from this implementation
+	NonInteger bool
+	Dimension  string
+}
+
+// Returns the Dimension field.
+func (e *TokenExtraNumeric) String() string {
+	if e == nil {
+		return ""
+	}
+	return e.Dimension
+}
+
+// TokenExtraUnicodeRange is attached to a TokenUnicodeRange.
+type TokenExtraUnicodeRange struct {
+	Start rune
+	End   rune
+}
+
+// Returns a valid CSS representation of the token.
+func (e *TokenExtraUnicodeRange) String() string {
+	if e == nil {
+		panic("TokenExtraUnicodeRange: unexpected nil pointer value")
+	}
+
+	if e.Start == e.End {
+		return fmt.Sprintf("U+%04X", e.Start)
+	} else {
+		return fmt.Sprintf("U+%04X-%04X", e.Start, e.End)
+	}
+}
+
+// TokenExtraError is attached to a TokenError and contains the same value as
+// Tokenizer.Err(). See also the ParseError type and ParseError.Recoverable().
+type TokenExtraError struct {
+	Err error
+}
+
+// Returns Err.Error().
+func (e *TokenExtraError) String() string {
+	return e.Err.Error()
+}
+
+// Error implements error.
+func (e *TokenExtraError) Error() string {
+	return e.Err.Error()
+}
+
+// Cause implements errors.Causer.
+func (e *TokenExtraError) Cause() error {
+	return e.Err
+}
+
+// Returns the ParseError object, if present.
+func (e *TokenExtraError) ParseError() *ParseError {
+	pe, ok := e.Err.(*ParseError)
+	if !ok {
+		return nil
+	}
+	return pe
+}
+
+func escapeIdentifier(s string) string { return escapeIdent(s, 0) }
+func escapeHashName(s string) string   { return escapeIdent(s, 1) }
+func escapeDimension(s string) string  { return escapeIdent(s, 2) }
+
+func needsHexEscaping(c byte, mode int) bool {
+	if c < 0x20 {
+		return true
+	}
+	if c >= utf8.RuneSelf {
+		return false
+	}
+	if mode == 2 {
+		if c == 'e' || c == 'E' {
+			return true
+		}
+	}
+	if c == '\\' {
+		return true
+	}
+	if isNameCode(c) {
+		return false
+	}
+	return true
+}
+
+func escapeIdent(s string, mode int) string {
+	if s == "" {
+		return ""
+	}
+	var buf bytes.Buffer
+	buf.Grow(len(s))
+	anyChanges := false
+
+	var i int
+
+	// Handle first character
+	// dashes allowed at start only for TokenIdent-ish
+	// eE not allowed at start for Dimension
```

Comment:
What was the reasoning behind this? It causes something like `asd {sdf: 3em}` to incorrectly turn into `asd {sdf: 3\65 m}` when re-encoding.

Reply:
I think this was referenced somewhere as being due to `1e9px` or something like that. Check the spec?